marketplaceItems/avatarCustomization101/appResources/appData/avatarCustomization101.js

```diff
@@ -0,0 +1,937 @@
+
+    // componentTypes
+    var STRING_COLOR = CONFIG.STRING_COLOR,
+        STRING_MAP_ONLY = CONFIG.STRING_MAP_ONLY;
+
+    var PATH_TO_IMAGES = MATERIAL_DATA.directory;
+
+    // propertyName
+    // newMaterialData { value: "" map: "" }
+    // componentType: STRING_COLOR / STRING_SLIDER / STRING_MAP_ONLY
+    // isPBR: boolean
+    function updateMaterialProperty(propertyName, newMaterialData, componentType, isPBR) {
+
+        var value = newMaterialData.value;
+        var map = newMaterialData.map;
+        
+        var property = propertyName;
+        var propertyMap = property + "Map";
+
+        if (DEBUG) {
+            print("Update Material Property key: ", property, " value: ", value);
+            print("Update Material Property key: ", propertyMap, " map: ", map);
+        }
+        
+        // update UI values
+        var type = isPBR ? "pbr" : "shadeless";
+
+        // prep data for changing material entity in updateMaterial()
+        var updates = {
+            unlit: isPBR ? false: true
+        };
+
+        if (value !== undefined) {
+            value = componentType === STRING_COLOR ? convertColorUIToBackend(value) : value;
+            value = componentType === STRING_MAP_ONLY ? convertImageUIToBackend(value) : value;
+            // slider value does not need to be changed
+
+            updates[property] = value;
+            dynamicData[STRING_MATERIAL][type][property].value = value;
+        }
+
+        if (map !== undefined && componentType !== STRING_MAP_ONLY) {
+            updates[propertyMap] = convertImageUIToBackend(map);
+            dynamicData[STRING_MATERIAL][type][property].map = map;
+        }
+
+        updateMaterial({ materials: updates }, false, isPBR);
+    }
+
+    function convertColorUIToBackend(hexColor){
+        // hex -> rgb -> array
+        var rgb = hexToRgb(hexColor);
+        return [ 
+            rgb.r / 255.0,
+            rgb.g / 255.0, 
+            rgb.b / 255.0 
+        ];
+    }
+
+    function convertImageUIToBackend(file) {
+        if (file && file.indexOf("no.jpg") !== -1) {
+            print("I AM HERE");
+            return null;
+        } 
+        return PATH_TO_IMAGES + file;
+    }
+
+
+    // prioritize a properties over b properties
+    function mergeObjectProperties(a, b) {
+
+        for(var key in b) {
+            a[key] = b[key];
+        }
+        return a;
+    }
+
+    function convertColorBackendToUI(arrayColor){
+        // array -> rgb -> hex
+        var rgb = arrayToRGB(arrayColor);
+        var hex = rgbToHex(rgb);
+        return hex;
+    }
+
+    // button is pressed and must loop through all UI elements to update
+    // newMaterialData.materials passed in
+    // for Named buttons only
+    function updateMaterialDynamicDataUI(newMaterialDataToApply, isPBR) {
+
+        // set the UI drop-down index to hifi-PBR (index 2) or shadeless (index 1)
+        dynamicData[STRING_MATERIAL].selectedTypeIndex = isPBR ? 2 : 1;
```

Comment:
are these indices guaranteed to be stable at 2 and 1 respectively? if not, consider making them CONSTs.